tfe/resource_tfe_variable.go

```diff
@@ -10,428 +10,709 @@ import (
+					stringplanmodifier.RequiresReplace(),
+				},
+				Validators: []validator.String{
+					stringvalidator.ExactlyOneOf(
+						path.MatchRelative().AtParent().AtName("variable_set_id"),
+					),
+					stringvalidator.RegexMatches(
+						workspaceIDRegexp,
+						"must be a valid workspace ID (ws-<RANDOM STRING>)",
+					),
+				},
 			},
-
-			"variable_set_id": {
-				Type:         schema.TypeString,
-				Optional:     true,
-				Required:     false,
-				ForceNew:     true,
-				ExactlyOneOf: []string{"workspace_id", "variable_set_id"},
-				ValidateFunc: validation.StringMatch(
-					variableSetIDRegexp,
-					"must be a valid variable set ID (varset-<RANDOM STRING>)",
-				),
+			"variable_set_id": schema.StringAttribute{
+				Optional: true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+				Validators: []validator.String{
+					stringvalidator.ExactlyOneOf(
+						path.MatchRelative().AtParent().AtName("workspace_id"),
+					),
+					stringvalidator.RegexMatches(
+						variableSetIDRegexp,
+						"must be a valid variable set ID (varset-<RANDOM STRING>)",
+					),
+				},
 			},
 		},
-
-		CustomizeDiff: forceRecreateResourceIf(),
+		Description:         "",
+		MarkdownDescription: "",
+		DeprecationMessage:  "",
+		Version:             1,
 	}
 }
 
-func forceRecreateResourceIf() schema.CustomizeDiffFunc {
-	/*
-		Destroy and add a new resource when:
-		1. the parameter key changed and the param sensitive is set to true
-		2. the parameter sensitive changed from true to false
-	*/
-	return func(ctx context.Context, d *schema.ResourceDiff, meta interface{}) error {
-		wasSensitiveVar := d.HasChange("sensitive") && !(d.Get("sensitive").(bool))
-
-		if wasSensitiveVar {
-			return d.ForceNew("sensitive")
-		} else if d.HasChange("key") && d.Get("sensitive").(bool) {
-			return d.ForceNew("key")
-		}
-		return nil
-	}
+// AttrGettable is a small enabler for helper functions that need to read one
+// attribute of a Plan or State.
+type AttrGettable interface {
+	GetAttribute(ctx context.Context, path path.Path, target interface{}) diag.Diagnostics
 }
 
-func resourceTFEVariableCreate(d *schema.ResourceData, meta interface{}) error {
-	// Switch to variable set variable logic if we need to
-	_, variableSetIDProvided := d.GetOk("variable_set_id")
-	if variableSetIDProvided {
-		return resourceTFEVariableSetVariableCreate(d, meta)
-	}
-
-	config := meta.(ConfiguredClient)
+// isWorkspaceVariable is a helper function for switching between tfe_variable's
+// two separate CRUD implementations.
+func isWorkspaceVariable(ctx context.Context, data AttrGettable) bool {
+	var variableSetID types.String
+	// We're ignoring the diagnostics returned by GetAttribute, because we'll
+	// be destructuring the entire schema value shortly in the real
+	// implementations; any notable problems will be reported at that point.
+	data.GetAttribute(ctx, path.Root("variable_set_id"), &variableSetID)
```

Comment:
I think it might be clearer if `isWorkspaceVariable` checks that workspaceID is present as opposed to the check for the absence of variableSetID. Is there a reason this route was chosen? WDYT?

Reply:
I think that was just a combination of my tendency to avoid a negated `if` when it's convenient, plus some influence from the logic in the old implementation. You're probably right, might as well switch it to a `if !workspace_id.IsNull()`.